src/integrationTest/java/com/mongodb/hibernate/EmbeddableIntegrationTests.java

```diff
@@ -0,0 +1,411 @@
+                .containsExactly(
+                        // Hibernate ORM does not store `item.omitted` despite it being non-`null`
+                        new BsonDocument(ID_FIELD_NAME, new BsonInt32(item.id)));
+        var loadedItem =
+                sessionFactoryScope.fromTransaction(session -> session.find(ItemWithFattenedEmptyValue.class, item.id));
+        // the entity we stored and the entity we loaded are not equal because Hibernate ORM omits `item.omitted`
+        assertNotEquals(item, loadedItem);
+        var updatedItem = sessionFactoryScope.fromTransaction(session -> {
+            var result = session.find(ItemWithFattenedEmptyValue.class, item.id);
+            result.omitted = null;
+            return result;
+        });
+        assertThat(mongoCollection.find()).containsExactly(new BsonDocument(ID_FIELD_NAME, new BsonInt32(item.id)));
+        loadedItem =
+                sessionFactoryScope.fromTransaction(session -> session.find(ItemWithFattenedEmptyValue.class, item.id));
+        assertEquals(updatedItem, loadedItem);
+    }
+
+    @Test
+    void testNestedValues() {
+        var item = new ItemWithNestedValues();
+        {
+            item.id = new StructAggregateEmbeddableValue();
+            item.id.a = 1;
+            item.nested1 = new StructAggregateEmbeddableValue();
+            item.nested1.a = 2;
+            item.nested2 = new StructAggregateEmbeddablePairValue1();
+            item.nested2.a = 3;
+            item.nested2.nested = new StructAggregateEmbeddablePairValue2(4, 5);
+            item.nested2.parent = item;
+        }
+        sessionFactoryScope.inTransaction(session -> session.persist(item));
+        assertThat(mongoCollection.find())
+                .containsExactly(new BsonDocument()
+                        // Hibernate ORM flattens `item.id` despite it being of an aggregate type
+                        .append(ID_FIELD_NAME, new BsonInt32(item.id.a))
+                        .append("nested1", new BsonDocument("a", new BsonInt32(2)))
+                        .append(
+                                "nested2",
+                                new BsonDocument()
+                                        .append("a", new BsonInt32(3))
+                                        .append(
+                                                "nested",
+                                                new BsonDocument()
+                                                        .append("a", new BsonInt32(4))
+                                                        .append("b", new BsonInt32(5)))));
+        var loadedItem =
+                sessionFactoryScope.fromTransaction(session -> session.find(ItemWithNestedValues.class, item.id));
+        assertEquals(item, loadedItem);
+        var updatedItem = sessionFactoryScope.fromTransaction(session -> {
+            var result = session.find(ItemWithNestedValues.class, item.id);
+            result.nested1.a = -result.nested1.a;
```

Comment:
Let's modify all possible values? For example, at least:
```
            result.nested1.a = -result.nested1.a;
            result.nested2.nested = new StructAggregateEmbeddablePairValue2(-4, -5);
```

I am unsure of how recursion works, so perhaps there is some way to miss deeply nested changes.

Reply:
We have all of the values "modified" when we insert. The only reason I test update is to have only some of them "modified".